File: src/main/java/org/bsplines/ltexls/parsing/CodeAnnotatedTextBuilder.java
package org.bsplines.ltexls.parsing;

import java.util.Map;
import java.util.HashMap;
import java.util.function.Function;

import org.bsplines.ltexls.Settings;
import org.bsplines.ltexls.Tools;
import org.bsplines.ltexls.parsing.latex.LatexAnnotatedTextBuilder;
import org.bsplines.ltexls.parsing.markdown.MarkdownAnnotatedTextBuilder;
import org.bsplines.ltexls.parsing.plaintext.PlaintextAnnotatedTextBuilder;
import org.checkerframework.checker.nullness.qual.Nullable;

import org.languagetool.markup.AnnotatedTextBuilder;

public abstract class CodeAnnotatedTextBuilder extends AnnotatedTextBuilder {
  private static Map<String, Function<String, CodeAnnotatedTextBuilder>> constructorMap =
      new HashMap<>();

  static {
    constructorMap.put("latex", (String codeLanguageId) ->
        new LatexAnnotatedTextBuilder(codeLanguageId));
    constructorMap.put("markdown", (String codeLanguageId) ->
        new MarkdownAnnotatedTextBuilder());
    constructorMap.put("plaintext", (String codeLanguageId) ->
        new PlaintextAnnotatedTextBuilder());
    constructorMap.put("rsweave", (String codeLanguageId) ->
        new LatexAnnotatedTextBuilder(codeLanguageId));
  }

  public static CodeAnnotatedTextBuilder create(String codeLanguageId) {
    @Nullable Function<String, CodeAnnotatedTextBuilder> constructor =
        constructorMap.get(codeLanguageId);

    if (constructor != null) {
      return constructor.apply(codeLanguageId);
    } else {
      Tools.logger.warning(Tools.i18n("invalidCodeLanguageId", codeLanguageId));
      return new PlaintextAnnotatedTextBuilder();
    }
  }

  public abstract CodeAnnotatedTextBuilder addCode(String code);

  public void setSettings(Settings settings) {}
}

File: src/main/kotlin/org/bsplines/ltexls/parsing/CodeAnnotatedTextBuilder.kt
/* Copyright (C) 2019-2023 Julian Valentin, LTeX Development Community
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package org.bsplines.ltexls.parsing

import org.bsplines.ltexls.parsing.gitcommit.GitCommitAnnotatedTextBuilder
import org.bsplines.ltexls.parsing.html.HtmlAnnotatedTextBuilder
import org.bsplines.ltexls.parsing.latex.LatexAnnotatedTextBuilder
import org.bsplines.ltexls.parsing.markdown.MarkdownAnnotatedTextBuilder
import org.bsplines.ltexls.parsing.nop.NopAnnotatedTextBuilder
import org.bsplines.ltexls.parsing.org.OrgAnnotatedTextBuilder
import org.bsplines.ltexls.parsing.plaintext.PlaintextAnnotatedTextBuilder
import org.bsplines.ltexls.parsing.program.ProgramAnnotatedTextBuilder
import org.bsplines.ltexls.parsing.program.ProgramCommentRegexs
import org.bsplines.ltexls.parsing.restructuredtext.RestructuredtextAnnotatedTextBuilder
import org.bsplines.ltexls.settings.Settings
import org.bsplines.ltexls.tools.I18n
import org.bsplines.ltexls.tools.Logging
import org.languagetool.markup.AnnotatedText
import org.languagetool.markup.AnnotatedTextBuilder
import org.languagetool.markup.TextPart

abstract class CodeAnnotatedTextBuilder(
  val codeLanguageId: String,
) : AnnotatedTextBuilder() {
  protected var curText = StringBuilder()
  protected var curMarkup = StringBuilder()
  protected var curInterpretAs = StringBuilder()
  protected var curType: TextPart.Type? = null

  abstract fun addCode(code: String): CodeAnnotatedTextBuilder

  @Suppress("UNUSED_PARAMETER")
  open fun setSettings(settings: Settings) {
  }

  override fun addText(text: String?): CodeAnnotatedTextBuilder {
    if (text?.isNotEmpty() == true) {
      if (curType == TextPart.Type.MARKUP) {
        finalizeCurrentPart()
      }
      curType = TextPart.Type.TEXT
      curText.append(text)
    }

    return this
  }

  override fun addMarkup(markup: String?): CodeAnnotatedTextBuilder {
    if (markup?.isNotEmpty() == true) {
      if (curType == TextPart.Type.TEXT) {
        finalizeCurrentPart()
      }
      curType = TextPart.Type.MARKUP
      curMarkup.append(markup)
    }

    return this
  }

  override fun addMarkup(markup: String?, interpretAs: String?): CodeAnnotatedTextBuilder {
    if (interpretAs?.isNotEmpty() == true) {
      if (curType == TextPart.Type.TEXT) {
        finalizeCurrentPart()
      }
      curType = TextPart.Type.MARKUP
      curMarkup.append(markup ?: "")
      curInterpretAs.append(interpretAs)
    } else {
      addMarkup(markup)
    }

    return this
  }

  override fun build(): AnnotatedText {
    finalizeCurrentPart()
    return super.build()
  }

  private fun finalizeCurrentPart() {
    if (curType == TextPart.Type.MARKUP) {
      super.addMarkup(curMarkup.toString(), curInterpretAs.toString())
      curMarkup.clear()
      curInterpretAs.clear()
    }
    if (curType == TextPart.Type.TEXT) {
      super.addText(curText.toString())
      curText.clear()
    }
  }

  companion object {
    @Suppress("ComplexMethod")
    fun create(codeLanguageId: String): CodeAnnotatedTextBuilder {
      return when (codeLanguageId) {
        "bib",
        "bibtex",
        -> LatexAnnotatedTextBuilder(codeLanguageId)
        "git-commit",
        "gitcommit",
        -> GitCommitAnnotatedTextBuilder(codeLanguageId)
        "html",
        "xhtml",
        -> HtmlAnnotatedTextBuilder(codeLanguageId)
        "context",
        "context.tex",
        "latex",
        "plaintex",
        "rsweave",
        "tex",
        -> LatexAnnotatedTextBuilder(codeLanguageId)
        "markdown",
        "quarto",
        "rmd",
        -> MarkdownAnnotatedTextBuilder(codeLanguageId)
        "nop" -> NopAnnotatedTextBuilder(codeLanguageId)
        "org" -> OrgAnnotatedTextBuilder(codeLanguageId)
        "plaintext" -> PlaintextAnnotatedTextBuilder(codeLanguageId)
        "restructuredtext" -> RestructuredtextAnnotatedTextBuilder(codeLanguageId)
        else -> {
          if (ProgramCommentRegexs.isSupportedCodeLanguageId(codeLanguageId)) {
            ProgramAnnotatedTextBuilder(codeLanguageId)
          } else {
            Logging.LOGGER.warning(I18n.format("unsupportedCodeLanguageId", codeLanguageId))
            PlaintextAnnotatedTextBuilder("plaintext")
          }
        }
      }
    }
  }
}
